### ISSUE TITLE
setup NS network: failed to add host route: file exists / UpdateRuleListBySrc: failed to add pod rule: file exists

### ISSUE THREAD

**Issue by tsuna** [aws/amazon-vpc-cni-k8s#344]
CNI version: v1.3.2
Kubernetes v1.11.8

On a small test cluster with 1 master (m3.medium) and 4 nodes (t2.small) running a total of 48 pods (across all namespaces, including `kube-system`) we frequently run into:
```
Mar 12 02:32:03 ip-172-20-44-167 kubelet[1952]: E0312 02:32:03.491035    1952 cni.go:260] Error adding network: add command: failed to setup network: setup NS network: failed to add host route: file exists
Mar 12 05:06:36 ip-172-20-44-167 kubelet[1952]: E0312 05:06:36.632374    1952 cni.go:260] Error adding network: add command: failed to setup network: UpdateRuleListBySrc: failed to add pod rule: file exists
```
We can manually unwedge nodes by deleting (a) host routes and (b) IP rules for IP addresses that were listed as unassigned by L-IPAM. For some reason, it appears to occasionally not clean up after itself, and doesn’t tolerate these entries already existing.

The problem seems to trigger very easily when running close to the maximum capacity (in terms of max number of pods due to the limitation on number of secondary IPs/ENI).

Similar issues: #166 #108 #203 

```
$ ip route show table all
default via 172.20.32.1 dev eth1 table 2 
172.20.32.1 dev eth1 table 2 scope link 
default via 172.20.32.1 dev eth2 table 3 
172.20.32.1 dev eth2 table 3 scope link 
default via 172.20.32.1 dev eth0 
172.17.0.0/16 dev docker0 proto kernel scope link src 172.17.0.1 linkdown 
172.20.32.0/19 dev eth0 proto kernel scope link src 172.20.33.232 
172.20.34.72 dev eni56651d0d5ae scope link 
172.20.38.99 dev eni67bd3e91236 scope link 
172.20.49.167 dev eni37b749f1d85 scope link 
172.20.51.87 dev enieefb111de6c scope link 
172.20.51.96 dev eni6e021c9d2b1 scope link 
172.20.53.173 dev enicdfefde8caf scope link 
172.20.60.61 dev eni32b8dad4bd9 scope link 
172.20.60.76 dev eniad24475670e scope link 
172.20.61.234 dev enibf3c0876f74 scope link 
broadcast 127.0.0.0 dev lo table local proto kernel scope link src 127.0.0.1 
local 127.0.0.0/8 dev lo table local proto kernel scope host src 127.0.0.1 
local 127.0.0.1 dev lo table local proto kernel scope host src 127.0.0.1 
broadcast 127.255.255.255 dev lo table local proto kernel scope link src 127.0.0.1 
broadcast 172.17.0.0 dev docker0 table local proto kernel scope link src 172.17.0.1 linkdown 
local 172.17.0.1 dev docker0 table local proto kernel scope host src 172.17.0.1 
broadcast 172.17.255.255 dev docker0 table local proto kernel scope link src 172.17.0.1 linkdown 
broadcast 172.20.32.0 dev eth0 table local proto kernel scope link src 172.20.33.232 
local 172.20.33.232 dev eth0 table local proto kernel scope host src 172.20.33.232 
broadcast 172.20.63.255 dev eth0 table local proto kernel scope link src 172.20.33.232 
fe80::/64 dev eth0 proto kernel metric 256  pref medium
fe80::/64 dev eth2 proto kernel metric 256  pref medium
fe80::/64 dev enie23ba1cf0e7 proto kernel metric 256  pref medium
fe80::/64 dev enic3677f3f4b5 proto kernel metric 256  pref medium
fe80::/64 dev eni0523767c4d5 proto kernel metric 256  pref medium
fe80::/64 dev eni3d508afbfa1 proto kernel metric 256  pref medium
fe80::/64 dev eni0ac45738e91 proto kernel metric 256  pref medium
fe80::/64 dev enibc4c1aee4a1 proto kernel metric 256  pref medium
fe80::/64 dev enibe5e9fb0c56 proto kernel metric 256  pref medium
fe80::/64 dev eni50602c1f067 proto kernel metric 256  pref medium
fe80::/64 dev enic9e98a886e8 proto kernel metric 256  pref medium
fe80::/64 dev eni593ad02caef proto kernel metric 256  pref medium
fe80::/64 dev eni56651d0d5ae proto kernel metric 256  pref medium
fe80::/64 dev eni6e021c9d2b1 proto kernel metric 256  pref medium
fe80::/64 dev eni67bd3e91236 proto kernel metric 256  pref medium
fe80::/64 dev eth1 proto kernel metric 256  pref medium
fe80::/64 dev enieefb111de6c proto kernel metric 256  pref medium
fe80::/64 dev eni37b749f1d85 proto kernel metric 256  pref medium
fe80::/64 dev eni32b8dad4bd9 proto kernel metric 256  pref medium
fe80::/64 dev eniad24475670e proto kernel metric 256  pref medium
fe80::/64 dev enibf3c0876f74 proto kernel metric 256  pref medium
fe80::/64 dev enicdfefde8caf proto kernel metric 256  pref medium
fe80::/64 dev eni428c1e7e7c8 proto kernel metric 256  pref medium
unreachable default dev lo proto kernel metric 4294967295  error -101 pref medium
local ::1 dev lo table local proto none metric 0  pref medium
local fe80::35:b5ff:fe2a:4d4c dev lo table local proto none metric 0  pref medium
local fe80::43:58ff:feea:4ecc dev lo table local proto none metric 0  pref medium
local fe80::d8:f1ff:feca:c122 dev lo table local proto none metric 0  pref medium
local fe80::8d1:e6ff:fe9d:fabb dev lo table local proto none metric 0  pref medium
local fe80::1831:f9ff:fecd:971b dev lo table local proto none metric 0  pref medium
local fe80::206c:a2ff:feb1:f4fc dev lo table local proto none metric 0  pref medium
local fe80::209a:6aff:fe41:8e12 dev lo table local proto none metric 0  pref medium
local fe80::2853:a8ff:fe5c:6720 dev lo table local proto none metric 0  pref medium
local fe80::2c33:dcff:feab:8d8e dev lo table local proto none metric 0  pref medium
local fe80::30df:2ff:fe10:a4a dev lo table local proto none metric 0  pref medium
local fe80::30f2:d2ff:febd:665d dev lo table local proto none metric 0  pref medium
local fe80::4894:29ff:fe0b:b9c3 dev lo table local proto none metric 0  pref medium
local fe80::58c1:56ff:fe4f:8124 dev lo table local proto none metric 0  pref medium
local fe80::6050:dbff:fe54:56c4 dev lo table local proto none metric 0  pref medium
local fe80::6402:f6ff:fe20:f956 dev lo table local proto none metric 0  pref medium
local fe80::6c2d:2aff:fee5:2a11 dev lo table local proto none metric 0  pref medium
local fe80::840c:4aff:fe59:a3e7 dev lo table local proto none metric 0  pref medium
local fe80::8417:a9ff:fe08:96fe dev lo table local proto none metric 0  pref medium
local fe80::84f2:2eff:fe13:a49 dev lo table local proto none metric 0  pref medium
local fe80::a881:b4ff:fedf:549e dev lo table local proto none metric 0  pref medium
local fe80::ac06:f6ff:fe75:20f0 dev lo table local proto none metric 0  pref medium
local fe80::d88a:9aff:fe9a:1c58 dev lo table local proto none metric 0  pref medium
local fe80::f8b0:55ff:fe51:89d6 dev lo table local proto none metric 0  pref medium
ff00::/8 dev eth0 table local metric 256  pref medium
ff00::/8 dev eth2 table local metric 256  pref medium
ff00::/8 dev enie23ba1cf0e7 table local metric 256  pref medium
ff00::/8 dev enic3677f3f4b5 table local metric 256  pref medium
ff00::/8 dev eni0523767c4d5 table local metric 256  pref medium
ff00::/8 dev eni3d508afbfa1 table local metric 256  pref medium
ff00::/8 dev eni0ac45738e91 table local metric 256  pref medium
ff00::/8 dev enibc4c1aee4a1 table local metric 256  pref medium
ff00::/8 dev enibe5e9fb0c56 table local metric 256  pref medium
ff00::/8 dev eni50602c1f067 table local metric 256  pref medium
ff00::/8 dev enic9e98a886e8 table local metric 256  pref medium
ff00::/8 dev eni593ad02caef table local metric 256  pref medium
ff00::/8 dev eni56651d0d5ae table local metric 256  pref medium
ff00::/8 dev eni6e021c9d2b1 table local metric 256  pref medium
ff00::/8 dev eni67bd3e91236 table local metric 256  pref medium
ff00::/8 dev eth1 table local metric 256  pref medium
ff00::/8 dev enieefb111de6c table local metric 256  pref medium
ff00::/8 dev eni37b749f1d85 table local metric 256  pref medium
ff00::/8 dev eni32b8dad4bd9 table local metric 256  pref medium
ff00::/8 dev eniad24475670e table local metric 256  pref medium
ff00::/8 dev enibf3c0876f74 table local metric 256  pref medium
ff00::/8 dev enicdfefde8caf table local metric 256  pref medium
ff00::/8 dev eni428c1e7e7c8 table local metric 256  pref medium
unreachable default dev lo proto kernel metric 4294967295  error -101 pref medium
```
```
$ kubectl -n kube-ingress describe pod ingress-nginx-75dc9cb46b-bzrmv
[...]
  Warning  FailedCreatePodSandBox  19m                  kubelet, ip-172-20-33-232.us-west-2.compute.internal  Failed create pod sandbox: rpc error: code = Unknown desc = failed to set up sandbox container "03a1aeac5f4850caf7dbf1e19a9cac0189851b916614729486eb66f6c81c76d5" network for pod "ingress-nginx-75dc9cb46b-bzrmv": NetworkPlugin cni failed to set up pod "ingress-nginx-75dc9cb46b-bzrmv_kube-ingress" network: add command: failed to setup network: setup NS network: failed to add host route: file exists
  Normal   SandboxChanged          16m (x21 over 21m)   kubelet, ip-172-20-33-232.us-west-2.compute.internal  Pod sandbox changed, it will be killed and re-created.
  Warning  FailedCreatePodSandBox  107s (x70 over 19m)  kubelet, ip-172-20-33-232.us-west-2.compute.internal  (combined from similar events): Failed create pod sandbox: rpc error: code = Unknown desc = failed to set up sandbox container "37711b390dfa2d1e6ff5877fcd598ba6a5559f5647d075fc0b34361659c16740" network for pod "ingress-nginx-75dc9cb46b-bzrmv": NetworkPlugin cni failed to set up pod "ingress-nginx-75dc9cb46b-bzrmv_kube-ingress" network: add cmd: failed to assign an IP address to container
```
Recent ipamd log:
```
2019-03-12T20:38:44Z [DEBUG] AssignIPv4Address: IP address pool stats: total:9, assigned 2
2019-03-12T20:38:44Z [INFO] DataStore has no available IP addresses
2019-03-12T20:38:44Z [DEBUG] VPC CIDR %!s(*string=0xc4202343d0)
2019-03-12T20:38:44Z [INFO] Send AddNetworkReply: IPv4Addr , DeviceNumber: 0, err: datastore: no available IP addresses
2019-03-12T20:38:44Z [INFO] Received DelNetwork for IP <nil>, Pod kube-state-metrics-864f6bdd88-rqmg5, Namespace monitoring, Container 83e77efd3ec62b54599d803329af0145668cc06992cc94b77ef9a8f38ebd787e
2019-03-12T20:38:44Z [DEBUG] UnAssignIPv4Address: IP address pool stats: total:9, assigned 2, pod(Name: kube-state-metrics-864f6bdd88-rqmg5, Namespace: monitoring, Container 83e77efd3ec62b54599d803329af0145668cc06992cc94b77ef9a8f38ebd787e)
2019-03-12T20:38:44Z [WARN] UnassignIPv4Address: Failed to find pod kube-state-metrics-864f6bdd88-rqmg5 namespace monitoring Container 83e77efd3ec62b54599d803329af0145668cc06992cc94b77ef9a8f38ebd787e
2019-03-12T20:38:44Z [DEBUG] UnAssignIPv4Address: IP address pool stats: total:9, assigned 2, pod(Name: kube-state-metrics-864f6bdd88-rqmg5, Namespace: monitoring, Container )
2019-03-12T20:38:44Z [WARN] UnassignIPv4Address: Failed to find pod kube-state-metrics-864f6bdd88-rqmg5 namespace monitoring Container 
2019-03-12T20:38:44Z [INFO] Send DelNetworkReply: IPv4Addr , DeviceNumber: 0, err: datastore: unknown pod
2019-03-12T20:38:44Z [INFO] Received AddNetwork for NS /proc/20075/ns/net, Pod ingress-nginx-75dc9cb46b-bzrmv, NameSpace kube-ingress, Container 81ef5ed3491460a9c679d75c23a42c42e1c97c18811328c997c62d32d2e69122, ifname eth0
2019-03-12T20:38:44Z [DEBUG] AssignIPv4Address: IP address pool stats: total:9, assigned 2
2019-03-12T20:38:44Z [INFO] AssignPodIPv4Address Assign IP 172.20.60.76 to pod (name ingress-nginx-75dc9cb46b-bzrmv, namespace kube-ingress container 81ef5ed3491460a9c679d75c23a42c42e1c97c18811328c997c62d32d2e69122)
2019-03-12T20:38:44Z [DEBUG] VPC CIDR %!s(*string=0xc4202343d0)
2019-03-12T20:38:44Z [INFO] Send AddNetworkReply: IPv4Addr 172.20.60.76, DeviceNumber: 3, err: <nil>
2019-03-12T20:38:44Z [DEBUG] IP pool stats: total = 9, used = 3, c.currentMaxAddrsPerENI = 3, c.maxAddrsPerENI = 3
2019-03-12T20:38:44Z [DEBUG] IP pool stats: total = 9, used = 3, c.currentMaxAddrsPerENI = 3, c.maxAddrsPerENI = 3
2019-03-12T20:38:44Z [DEBUG] No ENI can be deleted at this time
2019-03-12T20:38:44Z [ERROR] Failed to decrease pool free ENI: no ENI can be deleted at this time
2019-03-12T20:38:44Z [INFO] Received DelNetwork for IP 172.20.60.76, Pod ingress-nginx-75dc9cb46b-bzrmv, Namespace kube-ingress, Container 81ef5ed3491460a9c679d75c23a42c42e1c97c18811328c997c62d32d2e69122
2019-03-12T20:38:44Z [DEBUG] UnAssignIPv4Address: IP address pool stats: total:9, assigned 3, pod(Name: ingress-nginx-75dc9cb46b-bzrmv, Namespace: kube-ingress, Container 81ef5ed3491460a9c679d75c23a42c42e1c97c18811328c997c62d32d2e69122)
2019-03-12T20:38:44Z [INFO] UnAssignIPv4Address: pod (Name: ingress-nginx-75dc9cb46b-bzrmv, NameSpace kube-ingress Container 81ef5ed3491460a9c679d75c23a42c42e1c97c18811328c997c62d32d2e69122)'s ipAddr 172.20.60.76, DeviceNumber3
2019-03-12T20:38:44Z [INFO] Send DelNetworkReply: IPv4Addr 172.20.60.76, DeviceNumber: 3, err: <nil>
```
excerpt from `plugin.log`:
```
2019-03-12T20:40:42Z [INFO] Starting CNI Plugin v1.3.2  ...
2019-03-12T20:40:42Z [INFO] Received CNI del request: ContainerID(401bc7b65225585f61faa3c078f45c5abc7479f4f3fa442e74214a30c4f8a0e4) Netns() IfName(eth0) Args(IgnoreUnknown=1;K8S_POD_NAMESPACE=kube-ingress;K8S_POD_NAME=ingress-nginx-75dc9cb46b-bzrmv;K8S_POD_INFRA_CONTAINER_ID=401bc7b65225585f61faa3c078f45c5abc7479f4f3fa442e74214a30c4f8a0e4) Path(/opt/cni/bin/) argsStdinData({"cniVersion":"","name":"aws-cni","type":"aws-cni","vethPrefix":"eni"})
2019-03-12T20:40:42Z [ERROR] Failed to process delete request for pod ingress-nginx-75dc9cb46b-bzrmv namespace kube-ingress container 401bc7b65225585f61faa3c078f45c5abc7479f4f3fa442e74214a30c4f8a0e4: <nil>
[...]
2019-03-12T20:40:43Z [INFO] Starting CNI Plugin v1.3.2  ...
2019-03-12T20:40:43Z [INFO] Received CNI add request: ContainerID(4537a11399ab7a775d9714dc00674e8957f82cf78399db1fb3994c128a71811f) Netns(/proc/1361/ns/net) IfName(eth0) Args(IgnoreUnknown=1;K8S_POD_NAMESPACE=kube-ingress;K8S_POD_NAME=ingress-nginx-75dc9cb46b-bzrmv;K8S_POD_INFRA_CONTAINER_ID=4537a11399ab7a775d9714dc00674e8957f82cf78399db1fb3994c128a71811f) Path(/opt/cni/bin/) argsStdinData({"cniVersion":"","name":"aws-cni","type":"aws-cni","vethPrefix":"eni"})
[...]
2019-03-12T20:40:44Z [INFO] Starting CNI Plugin v1.3.2  ...
2019-03-12T20:40:44Z [INFO] Received CNI del request: ContainerID(4537a11399ab7a775d9714dc00674e8957f82cf78399db1fb3994c128a71811f) Netns(/proc/1361/ns/net) IfName(eth0) Args(IgnoreUnknown=1;K8S_POD_NAMESPACE=kube-ingress;K8S_POD_NAME=ingress-nginx-75dc9cb46b-bzrmv;K8S_POD_INFRA_CONTAINER_ID=4537a11399ab7a775d9714dc00674e8957f82cf78399db1fb3994c128a71811f) Path(/opt/cni/bin/) argsStdinData({"cniVersion":"","name":"aws-cni","type":"aws-cni","vethPrefix":"eni"})
2019-03-12T20:40:44Z [ERROR] Failed to process delete request for pod ingress-nginx-75dc9cb46b-bzrmv namespace kube-ingress container 4537a11399ab7a775d9714dc00674e8957f82cf78399db1fb3994c128a71811f: <nil>
```
Dump of ENIs:
```
$ curl -s http://localhost:61678/v1/enis | python -m json.tool
{
    "AssignedIPs": 2,
    "ENIIPPools": {
        "eni-09d41815c7d3e2419": {
            "AssignedIPv4Addresses": 1,
            "DeviceNumber": 2,
            "ID": "eni-09d41815c7d3e2419",
            "IPv4Addresses": {
                "172.20.49.167": {
                    "Assigned": false
                },
                "172.20.51.87": {
                    "Assigned": false
                },
                "172.20.61.234": {
                    "Assigned": true
                }
            },
            "IsPrimary": false
        },
        "eni-0a4db9cc2d9088b80": {
            "AssignedIPv4Addresses": 0,
            "DeviceNumber": 3,
            "ID": "eni-0a4db9cc2d9088b80",
            "IPv4Addresses": {
                "172.20.34.72": {
                    "Assigned": false
                },
                "172.20.60.61": {
                    "Assigned": false
                },
                "172.20.60.76": {
                    "Assigned": false
                }
            },
            "IsPrimary": false
        },
        "eni-0a80d51f06da74e87": {
            "AssignedIPv4Addresses": 1,
            "DeviceNumber": 0,
            "ID": "eni-0a80d51f06da74e87",
            "IPv4Addresses": {
                "172.20.38.99": {
                    "Assigned": false
                },
                "172.20.51.96": {
                    "Assigned": false
                },
                "172.20.53.173": {
                    "Assigned": true
                }
            },
            "IsPrimary": true
        }
    },
    "TotalIPs": 9
}
```
Metrics
```
$ curl -s http://localhost:61678/metrics
# HELP add_ip_req_count The number of add IP address request
# TYPE add_ip_req_count counter
add_ip_req_count 1411
# HELP assigned_ip_addresses The number of IP addresses assigned to pods
# TYPE assigned_ip_addresses gauge
assigned_ip_addresses 2
# HELP del_ip_req_count The number of delete IP address request
# TYPE del_ip_req_count counter
del_ip_req_count{reason="PodDeleted"} 2912
del_ip_req_count{reason="SetupNSFailed"} 688
# HELP eni_allocated The number of ENIs allocated
# TYPE eni_allocated gauge
eni_allocated 3
# HELP eni_max The maximum number of ENIs that can be attached to the instance
# TYPE eni_max gauge
eni_max 3
# HELP go_gc_duration_seconds A summary of the GC invocation durations.
# TYPE go_gc_duration_seconds summary
go_gc_duration_seconds{quantile="0"} 1.2327e-05
go_gc_duration_seconds{quantile="0.25"} 1.749e-05
go_gc_duration_seconds{quantile="0.5"} 1.8456e-05
go_gc_duration_seconds{quantile="0.75"} 2.019e-05
go_gc_duration_seconds{quantile="1"} 0.019268791
go_gc_duration_seconds_sum 0.023269378
go_gc_duration_seconds_count 125
# HELP go_goroutines Number of goroutines that currently exist.
# TYPE go_goroutines gauge
go_goroutines 25
# HELP go_memstats_alloc_bytes Number of bytes allocated and still in use.
# TYPE go_memstats_alloc_bytes gauge
go_memstats_alloc_bytes 6.861688e+06
# HELP go_memstats_alloc_bytes_total Total number of bytes allocated, even if freed.
# TYPE go_memstats_alloc_bytes_total counter
go_memstats_alloc_bytes_total 5.49008128e+08
# HELP go_memstats_buck_hash_sys_bytes Number of bytes used by the profiling bucket hash table.
# TYPE go_memstats_buck_hash_sys_bytes gauge
go_memstats_buck_hash_sys_bytes 1.483244e+06
# HELP go_memstats_frees_total Total number of frees.
# TYPE go_memstats_frees_total counter
go_memstats_frees_total 1.867272e+06
# HELP go_memstats_gc_sys_bytes Number of bytes used for garbage collection system metadata.
# TYPE go_memstats_gc_sys_bytes gauge
go_memstats_gc_sys_bytes 614400
# HELP go_memstats_heap_alloc_bytes Number of heap bytes allocated and still in use.
# TYPE go_memstats_heap_alloc_bytes gauge
go_memstats_heap_alloc_bytes 6.861688e+06
# HELP go_memstats_heap_idle_bytes Number of heap bytes waiting to be used.
# TYPE go_memstats_heap_idle_bytes gauge
go_memstats_heap_idle_bytes 2.58048e+06
# HELP go_memstats_heap_inuse_bytes Number of heap bytes that are in use.
# TYPE go_memstats_heap_inuse_bytes gauge
go_memstats_heap_inuse_bytes 9.609216e+06
# HELP go_memstats_heap_objects Number of allocated objects.
# TYPE go_memstats_heap_objects gauge
go_memstats_heap_objects 43433
# HELP go_memstats_heap_released_bytes_total Total number of heap bytes released to OS.
# TYPE go_memstats_heap_released_bytes_total counter
go_memstats_heap_released_bytes_total 0
# HELP go_memstats_heap_sys_bytes Number of heap bytes obtained from system.
# TYPE go_memstats_heap_sys_bytes gauge
go_memstats_heap_sys_bytes 1.2189696e+07
# HELP go_memstats_last_gc_time_seconds Number of seconds since 1970 of last garbage collection.
# TYPE go_memstats_last_gc_time_seconds gauge
go_memstats_last_gc_time_seconds 1.5524233501014578e+09
# HELP go_memstats_lookups_total Total number of pointer lookups.
# TYPE go_memstats_lookups_total counter
go_memstats_lookups_total 14764
# HELP go_memstats_mallocs_total Total number of mallocs.
# TYPE go_memstats_mallocs_total counter
go_memstats_mallocs_total 1.910705e+06
# HELP go_memstats_mcache_inuse_bytes Number of bytes in use by mcache structures.
# TYPE go_memstats_mcache_inuse_bytes gauge
go_memstats_mcache_inuse_bytes 1736
# HELP go_memstats_mcache_sys_bytes Number of bytes used for mcache structures obtained from system.
# TYPE go_memstats_mcache_sys_bytes gauge
go_memstats_mcache_sys_bytes 16384
# HELP go_memstats_mspan_inuse_bytes Number of bytes in use by mspan structures.
# TYPE go_memstats_mspan_inuse_bytes gauge
go_memstats_mspan_inuse_bytes 120232
# HELP go_memstats_mspan_sys_bytes Number of bytes used for mspan structures obtained from system.
# TYPE go_memstats_mspan_sys_bytes gauge
go_memstats_mspan_sys_bytes 147456
# HELP go_memstats_next_gc_bytes Number of heap bytes when next garbage collection will take place.
# TYPE go_memstats_next_gc_bytes gauge
go_memstats_next_gc_bytes 9.8276e+06
# HELP go_memstats_other_sys_bytes Number of bytes used for other system allocations.
# TYPE go_memstats_other_sys_bytes gauge
go_memstats_other_sys_bytes 370444
# HELP go_memstats_stack_inuse_bytes Number of bytes in use by the stack allocator.
# TYPE go_memstats_stack_inuse_bytes gauge
go_memstats_stack_inuse_bytes 393216
# HELP go_memstats_stack_sys_bytes Number of bytes obtained from system for stack allocator.
# TYPE go_memstats_stack_sys_bytes gauge
go_memstats_stack_sys_bytes 393216
# HELP go_memstats_sys_bytes Number of bytes obtained by system. Sum of all system allocations.
# TYPE go_memstats_sys_bytes gauge
go_memstats_sys_bytes 1.521484e+07
# HELP ip_max The maximum number of IP addresses that can be allocated to the instance
# TYPE ip_max gauge
ip_max 9
# HELP ipamd_action_inprogress The number of ipamd actions inprogress
# TYPE ipamd_action_inprogress gauge
ipamd_action_inprogress{fn="decreaseIPPool"} 0
ipamd_action_inprogress{fn="nodeIPPoolReconcile"} 0
ipamd_action_inprogress{fn="nodeInit"} 0
ipamd_action_inprogress{fn="retryAllocENIIP"} 0
# HELP ipamd_error_count The number of errors encountered in ipamd
# TYPE ipamd_error_count counter
ipamd_error_count{error="free ENI: no ENI can be deleted at this time",fn="decreaseIPPoolFreeENIFailed"} 678
# HELP process_cpu_seconds_total Total user and system CPU time spent in seconds.
# TYPE process_cpu_seconds_total counter
process_cpu_seconds_total 19.16
# HELP process_max_fds Maximum number of open file descriptors.
# TYPE process_max_fds gauge
process_max_fds 1.048576e+06
# HELP process_open_fds Number of open file descriptors.
# TYPE process_open_fds gauge
process_open_fds 10
# HELP process_resident_memory_bytes Resident memory size in bytes.
# TYPE process_resident_memory_bytes gauge
process_resident_memory_bytes 3.473408e+07
# HELP process_start_time_seconds Start time of the process since unix epoch in seconds.
# TYPE process_start_time_seconds gauge
process_start_time_seconds 1.5524199656e+09
# HELP process_virtual_memory_bytes Virtual memory size in bytes.
# TYPE process_virtual_memory_bytes gauge
process_virtual_memory_bytes 4.7812608e+07
# HELP total_ip_addresses The total number of IP addresses
# TYPE total_ip_addresses gauge
total_ip_addresses 9
```
Let me know if I can provide any other useful logs/data.

**Comment by tsuna** [comment]
Ping? :)

**Comment by mogren** [comment]
Hi, thanks for filing the issue! 

I haven't had time debug this yet, but like you mentioned it's similar to several open issues. It's interesting that your `eni-0a4db9cc2d9088b80` has no assigned IPs, but still has not been freed. 
Will investigate this more once I've managed to resolve some of older open issues.

**Comment by tsuna** [comment]
This happens very consistently on a small test cluster of t2.small instances where the number of pods is close to the limit (of 11) and the instances are sometimes running a bit hot due to uneven balancing of pods. Makes the cluster unusable. I haven't yet seen this bug on bigger clusters of m5.large instances.

You should be able to reproduce this bug easily but I'm also happy to provide any other information you need. I already rolled back the CNI change but I can switch back to it relatively easily and usually the bug hits within a few hours.

**Comment by erolosty** [comment]
Seeing this issue too

**Comment by tsuna** [comment]
Just had this hit on a m5.large instance. Definitely doesn't happen as frequently there but once a node gets a little too loaded, it does. I don't know how people and EKS can use this plugin, it's so unstable it's unusable to us. I'm rolling it back everywhere.

**Comment by flegler** [comment]
Hitting this too on EKS1.11.9 with AMI amazon-eks-node-1.11-v20190329 (ami-0e82e73403dd69fa3) and cni 1.3.2 on instance type m4.2xlarge
```
$ kubectl describe daemonset aws-node --namespace kube-system | grep Image | cut -d "/" -f 2
amazon-k8s-cni:v1.3.2
```

**Comment by tiffanyfay** [comment]
This should be fixed by cni v1.4.0 with PR #367.
